### ISSUE TITLE
ERC865: Pay transfers in tokens instead of gas, in one transaction

### ISSUE THREAD

**Issue by lgalabru** [ethereum/EIPs#865]
## Preamble

	EIP: <to be assigned>
	Title: Token standard
	Author: Ludovic Galabru
	Type: Informational
	Category: ERC
	Status: Draft
	Created: 2018-01-30
	Requires: EIP20

## Simple Summary
Ability for token holders to pay transfer transactions in tokens instead of gas, in one transaction.

## Abstract
The following describes one standard function a token contract can implement to allow a user to delegate transfer of tokens to a third party. The third party pays for the gas, and takes a fee in tokens.

## Motivation
When it comes to using tokens as utility tokens, we need to strive for a good UX. Introducing wallets and transactions to end users is a challenge, and having to explain that token holders needs ETH to send tokens is adding some friction to the process. The goal of this EIP is to abstract the gas for the end user, by introducing a fee paid in tokens. A third party can then bring the transaction on-chain, pay for the gas of that given transaction and get tokens from that user.

## Specification
- A: Sender of the payment
- B: Recipient of the payment
- D: Delegate, doing the transaction for A, and paying for the gas.
- X: Amount of Token T sent from A to B
- Y: Fee paid in Token T, from A to D for the transaction
- T: Token to send
- N: Nonce

### Process

The user **A** gets a quote from the delegate **D** for the value of the fee **Y** for 1 transaction (depending on gas price + value of token in ETH).

With their private key, the user generates **{V,R,S}** for the **sha3** of the payload **P** {N,A,B,D,X,Y,T}.

The user sends **{V,R,S}** and **P** (unhashed, unsigned) to the delegate.

The delegate verifies that **Y** and **D** have not been altered. 

The delegate proceeds to submit the transaction from his account **D**:

    T.delegatedTransfer(N,A,B,X,Y,V,R,S)

The *delegatedTransfer* method reconstructs the sha3 **H** of the payload **P** (where **T** is the address of the current contract and **D** is the *msg.sender*).

We can then call *ecrecover(H, V, R, S)*, make sure that the result matches **A**, and if that’s the case, safely move **X** tokens from **A** to **B** and **Y** tokens from **A** to **D**.

The challenge mainly resides in imitating the *Non-standard Packed Mode* on the client side, and obtaining the exact same sha3 as the one generated on-chain.

### Methods

**delegatedTransfer**

    function transferPreSigned(
        bytes _signature,
        address _to,
        uint256 _value,
        uint256 _fee,
        uint256 _nonce
    )
        public
        returns (bool);

Is called by the delegate, and performs the transfer.

### Events

**TransferPreSigned**

    event TransferPreSigned(address indexed from, address indexed to, address indexed delegate, uint256 amount, uint256 fee);

Is triggered whenever delegatedTransfer is successfully called.

## Implementation proposal

Assuming a StandardToken and SafeMath available, one could come up with the following implementation.

#### On-chain operation (solidity)

```solidity
    /**
     * @notice Submit a presigned transfer
     * @param _signature bytes The signature, issued by the owner.
     * @param _to address The address which you want to transfer to.
     * @param _value uint256 The amount of tokens to be transferred.
     * @param _fee uint256 The amount of tokens paid to msg.sender, by the owner.
     * @param _nonce uint256 Presigned transaction number. Should be unique, per user.
     */
    function transferPreSigned(
        bytes _signature,
        address _to,
        uint256 _value,
        uint256 _fee,
        uint256 _nonce
    )
        public
        returns (bool)
    {
        require(_to != address(0));

        bytes32 hashedParams = transferPreSignedHashing(address(this), _to, _value, _fee, _nonce);
        address from = recover(hashedParams, _signature);
        require(from != address(0));

        bytes32 hashedTx = keccak256(from, hashedParams);
        require(hashedTxs[hashedTx] == false);

        balances[from] = balances[from].sub(_value).sub(_fee);
        balances[_to] = balances[_to].add(_value);
        balances[msg.sender] = balances[msg.sender].add(_fee);
        hashedTxs[hashedTx] = true;

        Transfer(from, _to, _value);
        Transfer(from, msg.sender, _fee);
        TransferPreSigned(from, _to, msg.sender, _value, _fee);
        return true;
    }

    /**
     * @notice Hash (keccak256) of the payload used by transferPreSigned
     * @param _token address The address of the token.
     * @param _to address The address which you want to transfer to.
     * @param _value uint256 The amount of tokens to be transferred.
     * @param _fee uint256 The amount of tokens paid to msg.sender, by the owner.
     * @param _nonce uint256 Presigned transaction number.
     */
    function transferPreSignedHashing(
        address _token,
        address _to,
        uint256 _value,
        uint256 _fee,
        uint256 _nonce
    )
        public
        pure
        returns (bytes32)
    {
        /* "48664c16": transferPreSignedHashing(address,address,address,uint256,uint256,uint256) */
        return keccak256(bytes4(0x48664c16), _token, _to, _value, _fee, _nonce);
    }

```

#### Off-chain usage (js)

```javascript

    describe(`if Charlie performs a transaction T, transfering 100 tokens from Alice to Bob (fee=10)`, () => {
      beforeEach(async () => {
        const nonce = 32;
        const from = alice;
        const to = bob;
        const delegate = charlie;
        const fee = 10;
        const amount = 100;
        const alicePrivateKey = Buffer.from('c87509a1c067bbde78beb793e6fa76530b6382a4c0241e5e4a9ec0a0f44dc0d3', 'hex');

        const components = [
          Buffer.from('48664c16', 'hex'),
          formattedAddress(this.token.address),
          formattedAddress(to),
          formattedInt(amount),
          formattedInt(fee),
          formattedInt(nonce)
        ];
        const vrs = ethUtil.ecsign(hashedTightPacked(components), alicePrivateKey);
        const sig = ethUtil.toRpcSig(vrs.v, vrs.r, vrs.s);
        await this.token.transferPreSigned(
          sig,
          to,
          amount,
          fee,
          nonce
          , {from: charlie}).should.be.fulfilled;
      });

```

## Full implementation available
OpenZeppelin/zeppelin-solidity#741

## Additional documentation
[Transfer Ethereum tokens without Ether — An ERC20 Improvement to Seriously Consider](https://medium.com/@adilharis/transfer-ethereum-tokens-without-ether-an-erc20-improvement-to-seriously-consider-90bebd447bb)

## Copyright
Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).

**Comment by WalterBoles** [comment]
All of translations where heals up by the blockchain or Coinbase so I will pay again but I would like to be assured that they will sell them and pay me

**Comment by 3esmit** [comment]
I'm working on something simmilar for about a week now, but I've come with a different implementation.
I've realized this idea while writing this comment https://ethresear.ch/t/pos-and-economic-abstraction-stakers-would-be-able-to-accept-gas-price-in-any-erc20-token/721/7?u=3esmit as I was told it was probably not going to happen I came up with this solution https://github.com/status-im/ideas/issues/73
A first version was deployed here: https://ropsten.etherscan.io/address/0x9787564e1bd7da95ee9dcdf17cc57a7225084632
And the latest version is available here https://github.com/status-im/contracts/blob/presigned-token/contracts/token/MiniMeTokenPreSigned.sol

**Comment by ptrwtts** [comment]
Nice work @3esmit. Do you have any comments on the pros / cons of this implementation vs yours?

**Comment by lgalabru** [comment]
@3esmit pretty cool!
I updated the code and added an exemple of the off chain implementation.

**Comment by bokkypoobah** [comment]
See https://github.com/bokkypoobah/BokkyPooBahsTokenTeleportationServiceSmartContract#how-it-works

**Comment by 3esmit** [comment]
@ptrwtts My approch support arbitrary contract executions in SNT network, so we can even create a new contract and pay the gas relayed by SNT "gas".

@lgalabru I'm not sure why supporting any contract address for building the txHash, initially I did so and used `address(this)` at it's calling, and should not be a problem. If you prefer that, I suggest setting function as pure. 

@bokkypoobah live in which network? Mainnet? 
I deployed an example of SNTPreSigned at Ropsten, check it out https://ropsten.etherscan.io/address/0xb473fd6c1206655bf2385013b23589f713fe3213

**Comment by bokkypoobah** [comment]
BTTSTokenFactory on mainnet - https://etherscan.io/address/0x594dd662b580ca58b1186af45551f34312e91e88#code

And while developing it, I was getting an out of stack space for variables problem in Solidity when using r,s and v, so switched to using sig.

**Comment by lgalabru** [comment]
@3esmit the idea was to use the same method (indeed, pure), on-chain and off chain. 
From the smart contract, you could have use:
```solidity
 bytes32 hash = getTransferHash(
        address(this)
        _to,
        _value,
        _gasPrice,
        _nonce
    )
```

And off-chain, you could have use:
```javascript
const payload = await token.getTransferHash.request(_contract,  _to, _value, _gasPrice, _nonce);
const data = payload.params[0].data;
const hash = ethUtil.sha3(data);
```

After testing this approach, the problem is that off chain, when building data, all the arguments are padded vs being tightly-packed on chain.

**Comment by 3esmit** [comment]
Fundamentally to process the contract call you need to be synced with state, the only difference is that you can call that function `pure` then, there is no other benefict, I would stay with being processed internally because this is a known common header of signature and to benefict of a simplier function signature. 
Maybe if gets cheaper in gas by using the pure function, then would be interesting, but I don't think this will make a difference overall.

**Comment by sponnet** [comment]
https://github.com/swarmcity/SCLabs-gasstation-service and https://github.com/swarmcity/SCLabs-gasstation-client

**Comment by seweso** [comment]
Wouldn't a way simpler solution be to have contracts set gasprice & startgas  so miners are allowed to take gas from the contracts balance? 

Then nodes/miners would use limited gas per transaction to determine if these transactions are mineable at all (gas price is high enough) and warrant to be re-broadcast over the network. In other words, nodes/miners are executing the code of transactions (where the sender doesn't have enough funds themselves), and which contain 'gasprice & startgas'-opcodes. 

Contracts can then convert tokens to gas themselves. Basically wallets would allow users to set gasprice & startgas in the used token, and the contract converts that to ETH. Or if the contract is rather predictable in gas usage make it even easier for users by allowing users to only set gasprice.

**Comment by aakilfernandes** [comment]
Very cool. This would be extremely useful from a UX perspective (users don't have to hold multiple tokens).

I'd also like to see a way to pay the delegate a fee (in the token, not ether). That way you could imagine a bunch of nodes which accept any delegate transaction in exchange for a fee.

**Comment by lgalabru** [comment]
@seweso I think the change you're describing needs to happen on the EVM level (and is on the roadmap if my understanding is correct). 
This is less trivial than just iterating on top of ERC20.

**Comment by mattdf** [comment]
It would be good to standardize this to cover delegated ETH transfers as well when token = 0x0, for things like withdrawing from ring signature mixers. As otherwise, you would have to "pre-fund" the withdrawing account, thereby linking your deposit and withdraw addresses.

**Comment by seweso** [comment]
@lgalabru Yes agreed. And it is a direction Ethereum needs to go into. 

Do you know how this is called on the roadmap? I couldn't find it.

**Comment by ojanssens** [comment]
I think this is a great idea and an answer to my problem I’ve been struggling with: https://forum.ethereum.org/discussion/16990/transaction-fees-for-an-erc20-currency

I only have 1 concern though: Wouldn’t this make Ether itself redundant? If miners would be able to directly accept tokens as payment, the utility of Ether would become less relevant except for staking and securing the network (which is still extremely valuable of course).

**Comment by Philipinho** [comment]
@bokkypoobah, i'm curious to ask. What happens if the tokens are without worthless? How will the BTTS pay for the gas in ether when the  token has no value?

**Comment by bokkypoobah** [comment]
@Philipinho I plan to provide a smart contract that the token contract owner has to top up with ETH, and this smart contract will buy tokens from the BTTS service provider at a specified rate.

**Comment by 3esmit** [comment]
I think that defining a gasPrice instead a fee is more safe and dynamic, there is no reason for not doing this.
It's trivial to calculate the gas cost of an operation inside the smart contract. The global `msg.gas` returns the remaining gas, so simply read this value at bounds of operation (right after sending the gas). 

This is really important for approveAndCallPreSigned, because this function can call arbritary execution in other contract that can be hard to be estimated, so we place the responsability of being safe to signer.
gasLimit could be interesting to safety of both sides, but the limiting already exists in native gas. 


Please review my work in [MiniMeTokenPreSigned.sol](https://github.com/status-im/contracts/blob/presigned-token/contracts/token/MiniMeTokenPreSigned.sol) and derive from it, it's the same GNU license followed by [MiniMeToken.sol](https://github.com/Giveth/minime/blob/master/contracts/MiniMeToken.sol) (I guess GPLv3, @jbaylina?). 

The interface I suggest is the following:
```
pragma solidity ^0.4.17;

/**
   @notice Implements PreSigned ERC20Token operations (and approveAndCall(address,uint256,bytes);
 */
contract ERC865 {
    /**
     * @notice Include a presigned `"a9059cbb": "transfer(address,uint256)"`
     * @param _signature Signed transfer 
     * @param _to The address of the recipient
     * @param _value The value of tokens to be transferred
     * @param _gasPrice How much tokens willing to pay per gas
     * @param _nonce Presigned transaction number.
     */
    function transferPreSigned(bytes _signature, address _to, uint256 _value, uint256 _gasPrice, uint256 _nonce) public;

    /**
     * @notice Include a presigned `""095ea7b3": "approve(address,uint256)"`
     * @param _signature Signed transfer 
     * @param _to The address of the recipient
     * @param _value The value of tokens to be transferred
     * @param _gasPrice How much tokens willing to pay per gas
     * @param _nonce Presigned transaction number.
     */    
    function approvePreSigned(bytes _signature, address _to, uint256 _value, uint256 _gasPrice, uint256 _nonce) public;
    
    /**
     * @notice Include a presigned `"cae9ca51": "approveAndCall(address,uint256,bytes)"`
     * @param _signature Signed transfer 
     * @param _to The address of the recipient
     * @param _value The value of tokens to be transferred
     * @param _extraData option data to send to contract
     * @param _gasPrice How much tokens willing to pay per gas
     * @param _nonce Presigned transaction number.
     */
    function approveAndCallPreSigned(bytes _signature, address _to, uint256 _value, uint256 _extraData, uint256 _gasPrice, uint256 _nonce) public;
    
}
```

**Comment by Arachnid** [comment]
Excellent idea! Why explicitly encode the sender when it's provided by ecrecover, though?

Also, why encode the delegate, instead of letting the first person to get a transaction in harvest the fee?

**Comment by jdkanani** [comment]
Great. @lgalabru `nonce` wouldn't work.  Possible race condition - `require(_nonce > nonces[_from])` prevents older transactions to go through, even if `sig` is valid.

**Comment by lgalabru** [comment]
@jdkanani 
Amazing, I had the race condition in my radar and was dubitative, thanks for your workaround!

@Arachnid 
Right, we need to pass it, but it doesn't have to be encoded. Good catch
> Letting the first person to get a transaction in harvest the fee
Great idea!

@3esmit 
Using msg.gas price makes sense, but we should probably come up with another name, since "gas price" is usually in wei. 
Concerning the naming of the methods, what would you think of prefixing the methods with "delegated", instead of suffixing with "presign", for insisting on the fact that someone else is bringing the transaction on chain? 
Concerning the direction, we probably want to open a pull request on Open Zeppelin (cc @spalladino). Would it work with the constraints of your licence?

**Comment by izqui** [comment]
I have been working on this idea for a side-project, the [pay-protocol](https://github.com/aragonlabs/pay-protocol/blob/master/contracts/PayProtocol.sol#L52). This feature is something that can be achieved in a **layer 2 protocol** rather than adding this functionality to every token contract. Which also has the ability that every token currently deployed could use it, just by having people that want to use this create a ERC20 allowance to the PayProtocol contract or setting it a ERC777 operator.

In any case, I think it would be super interesting to standardize the offchain part of this, which would allow people to run nodes looking for profitable token transfers and settle them.

**Comment by 3esmit** [comment]
@lgalabru "Delegated" means that you transfered temporarily trust to other, is not the case here, so I think the naming transferPreSigned() or signedTransfer() are more correct, but `transferPreSigned` actually describe exacly what the function is (a transfer which been pre signed.)

I don't think is bad to use `gasPrice` as the name, it also describe exactly what is this parameter for, and it works just like regular native ether gasPrice, however it is under the token you're interacting. 
Just like in native token, gasPrice is set in the minimum unit (wei), tokens that have decimals need to be aware of this. A good UX would handle this fine.

@izqui I think your project is a little bit different because is not the token itself that is handeling the moving, seems like is a TokenController? but should work like the same.
About the ethereum signed message, I see that you do this:
`keccak256(SCHEMA_HASH, keccak256(this, token, from, to, value, expires, pull, push, keccak256(deps)));`
Maybe we can have your opinion on why you did this and why we also should do something like this.

**Comment by axcrest** [comment]
How do you handle the ERC-20 approve() requirement?  No one can withdraw funds from your token account without you first approving them, and that approve() step costs ETH.

**Comment by izqui** [comment]
@axcrest you are right that in pay protocol if you already have tokens you need to create an approval to the contract. But if you receive the tokens directly in your pay protocol account, then you never need ether to operate.

@3esmit you are right the token isn't handling the moving and that's why it is very efficient, token transfers are only settled in the token contracts when tokens need to move in or out of the protocol. This allows to settle MiniMe transfers for ~15k gas (if using bulk transfers, ~36k otherwise) instead of +100k that takes if operating directly on the token.

The signed message conforms to the latest https://github.com/ethereum/EIPs/pull/712 spec. It allows for signing providers to show exactly what is going in the hash they are going to be signing.

**Comment by 3esmit** [comment]
@izqui Thanks for the information about signed message spec, I think the community should consensus in a default signed standard ASAP. 
I liked #721 and I think it should be a subtype (version) of #191. 
You are right about pay-protocol being cheaper, and it's an interesting solution, and should be used together with `approveAndCallPreSigned`. 
The first transaction from someone entering this network could be paid in the token itself directly. I think whats being discussed here and the network you showed us are different things and compliment each other. Users always want cheaper transactions, and also may want to not hold ether.

The main issue this EIP resolves is the need of an account to hold Ether in their balance to move their token. Even being more expansive, it might be cheaper then the process of transferring ether to that account and moving it. 

MiniMeTokenPresigned behaves differnt when gasPrice is zero, this can be used for an user which have 2 balances, 1 with ether and other with the token, so they can control this account using gas from other account. 

I can upgrade Raiden contracts to be able of accepting `approveAndCallPreSigned`  to open a channel, so users could participate in Raiden network without ever holding any ether. I think this is certainly an improvement for the token that reaches almost the same result as a true economic abstraction, with the restriction that contracts must be implemented `ApproveAndCallFallBack.sol` properly. 
I'll prepare examples in the week for StandardBounty contracts and fees being paid with `approveAndCallPresigned`.

**Comment by lgalabru** [comment]
A first implementation of this EIP is available here: https://github.com/PROPSProject/props-token-distribution/blob/master/contracts/token/ERC865Token.sol.
I'll probably open a pull request on zeppelin after getting some feedbacks from you guys.

@3esmit I'm not convinced by this idea of computing the gas cost on the fly because you're adding more complexity to the method, when you want to be really accurate.

I skipped the approveAndCall, since it's not part of the ERC20 protocol + there is no clear settlement at this point (ERC223 vs ERC827).

**Comment by ptrwtts** [comment]
@izqui I'm curious how you envisage a decentralized settlement system working for these sorts of transactions? 

We realized that if you leave the transactions open for anyone to settle, then it's possible for someone to watch the mempool and front run them.

You can avoid it, but there needs to be some back and forth. First, the user broadcasts their intent to make a transaction. Then delegates provide quotes. Then the user picks a delegate, and signs the transaction with their address included, so that no one else can front run.

In both cases there is a compromise, so not sure which is best.

**Comment by Arachnid** [comment]
I agree - being conformant with #191 makes much more sense to me.

**Comment by lgalabru** [comment]
@Arachnid what should I change on
https://github.com/PROPSProject/props-token-distribution/blob/master/contracts/token/ERC865Token.sol
if we want to conform to #191?

**Comment by AugustoL** [comment]
@lgalabru love the idea! please tag qhen you add the PR to zeppelin-solidity, looking forward to have it there!

**Comment by lgalabru** [comment]
@AugustoL PR available here: https://github.com/OpenZeppelin/zeppelin-solidity/pull/741

**Comment by 3esmit** [comment]
I'm working here with a contract that should be able to use approveAndCallPreSigned, https://github.com/status-im/StandardBounties/blob/bounty-manager/contracts/BountyManagerPreSigned.sol#L51

It's not ready yet, but thats the idea.

**Comment by Illasera** [comment]
Get ready for a billion tokens payment method,

Not only that you will scare the every-day layman, But you will scare the actual blockchain users
with over-saturated market. If it doesn't work that way in the real world, its a bad idea.

Imagine if you have will to pay in special coupons for every single thing.
Shirt store coupons only
Pants store coupons only
Burger-town coupons only.

**Comment by nakajo2011** [comment]
this is cool proposal! I think it is more useful to add that "bytes inputData" argument to a delegate transaction.

**Comment by 3esmit** [comment]
@Illasera If someone willing to pay ether-gas to get a valueless token then is problem of who is including the presigned transaction to the contract.

**Comment by SimonFricker** [comment]
What if tokens are created with an embedded GAS value. This way the GAS can be deducted from the token, and the token destroyed?

**Comment by 0xbitcoin** [comment]
This is fantastic.  The problem I see is that many tokens have been launched and will not add this functionality.  Furthermore, you do not need to pass in V R S you can just pass in bytes for the full signature and use ECRecover sample library from Zeppelin.  

In order to allow for older ERC20 tokens to have this functionality, I am building a special 'wallet' contract called LavaWallet.  This will allow all deposited tokens to be transferred to other parties via this same metholodogy, by relayers who turn in offchain signatures.   Unfortunately it will have to deviate from your standard since its a related but different concept.   

https://github.com/0xbitcoin/lava-wallet

**Comment by wadeAlexC** [comment]
Hey all! @RobertMCForster and I are working to implement Coinvest's COIN token to be ERC865 compatible. Obviously, that's difficult, seeing as ERC865 isn't finalized, but we're coming up with a few workarounds to (hopefully) have a forward-compatible system in place.

I think we've come up with a potential problem with the standard implemented as-is. For some context, the standard implementation of the ERC20 `approve` function is regarded by many as being vulnerable to a racing condition, wherein the following occurs:
```
1. Alice approves N tokens to Bob (N > 0)
2. Alice wants to approve M tokens to Bob instead (M > 0). She signs and broadcasts the transaction.
3. Bob notices the transaction, and quickly signs and broadcasts a transferFrom transaction with a much higher gas price.
4. Bob's transferFrom is picked up first by a miner, and Bob transfers N tokens from Alice's balance. Bob's allowance from Alice is now 0.
5. Alice's transaction is picked up by a miner, and Bob is now re-approved to the tune of M tokens.
6. Bob is now able to perform another transferFrom to transfer another M tokens from Alice.
```
@RobertMCForster and I feel that the ERC865 `approvePreSigned` and `approveAndCallPreSigned` functions are particularly vulnerable to this issue, given that if Alice provides some party P with a pre-signed `approve` or `approveAndCall` transaction, P now has a ticket to exploit the above racing condition without even needing to rely on a miner picking up their transaction first. If P is the party to which the tokens are approved, they can perform the exploit themselves. If not, they can sell the exploit to the affected party (or even just notify them of the pending approval). Regardless, it is now up to P to behave honestly.

Our proposed solution is two-fold: 

First, the functions `increaseApprovalPreSigned` and `decreaseApprovalPreSigned`, should be added to the standard as well. `approvePreSigned` and `approveAndCallPreSigned` do not necessarily need to be removed - we can leave it to the signer to decide if the interaction is safe or not.

Second, signers need an option to revoke a signed transaction, in the event they believe the broadcasting party will behave maliciously. Or, maybe they simply want to revoke the transaction without having to execute it themselves/increase the nonce with other transactions (depending on the implementation). We suggest the function `revokeSignature` as a standard function to address this scenario. The ERC865 interface would then be the following:

```Solidity
pragma solidity ^0.4.23;

/**
   @notice Implements PreSigned ERC20Token operations (and approveAndCall(address,uint256,bytes);
 */
interface ERC865 {
    /**
     * @notice Include a presigned `"a9059cbb": "transfer(address,uint256)"`
     * @param _signature Signed transfer 
     * @param _to The address of the recipient
     * @param _value The value of tokens to be transferred
     * @param _gasPrice How much tokens willing to pay per gas
     * @param _nonce Presigned transaction number.
     */
    function transferPreSigned(bytes _signature, address _to, uint256 _value, uint256 _gasPrice, uint256 _nonce) public;

    /**
     * @notice Include a presigned `"095ea7b3": "approve(address,uint256)"`
     * @param _signature Signed approval hash 
     * @param _spender The address of the spender
     * @param _value The amount of tokens to approve
     * @param _gasPrice How much tokens willing to pay per gas
     * @param _nonce Presigned transaction number.
     */    
    function approvePreSigned(bytes _signature, address _spender, uint256 _value, uint256 _gasPrice, uint256 _nonce) public;
    
    /**
     * @notice Include a presigned `"cae9ca51": "approveAndCall(address,uint256,bytes)"`
     * @param _signature Signed approval hash 
     * @param _spender The address of the spender
     * @param _value The amount of tokens to approve
     * @param _extraData option data to send to contract
     * @param _gasPrice How much tokens willing to pay per gas
     * @param _nonce Presigned transaction number.
     */
    function approveAndCallPreSigned(bytes _signature, address _spender, uint256 _value, bytes _extraData, uint256 _gasPrice, uint256 _nonce) public;

     /**
     * @notice Include a presigned `"d73dd623": "increaseApproval(address,uint256)"`
     * @param _signature Signed approval hash 
     * @param _spender The address of the spender
     * @param _amount The amount by which the spender's allowance will be increased
     * @param _gasPrice How much tokens willing to pay per gas
     * @param _nonce Presigned transaction number.
     */    
    function increaseApprovalPreSigned(bytes _signature, address _to, uint256 _amount, uint256 _gasPrice, uint256 _nonce) public;

    /**
     * @notice Include a presigned `"66188463": "decreaseApproval(address,uint256)"`
     * @param _signature Signed approval hash 
     * @param _spender The address of the spender
     * @param _amount The amount by which the spender's allowance will be decreased
     * @param _gasPrice How much tokens willing to pay per gas
     * @param _nonce Presigned transaction number.
     */    
    function decreaseApprovalPreSigned(bytes _signature, address _to, uint256 _amount, uint256 _gasPrice, uint256 _nonce) public;

    /**
     * @param _signature: Signature to be revoked by the sender
    */
    function revokeSignature(bytes _signature) public;
    
}
```

I would recommend one of the two following simple implementations for `revokeSignature`, but there are several possibilities:
```Solidity
/// Incrementing-nonce based:

// Maps a signer to their current pre-signed transaction nonce
mapping (address => uint256) public nonces;

// Functions only pass if the given signature has not been invalidated by the signer
modifier validSignature(address _signer, uint _nonce) {
    require(_nonce == nonces[_signer] + 1);
    _;
}    

/**
 * @param _signature: Signature to be revoked by the sender. In this case, the signature is irrelevant
*/
function revokeSignature(bytes _signature) public {
    _signature;
    nonces[msg.sender]++;
}
```
-or-
```Solidity
/// Non-incrementing-nonce based. We will likely use this to avoid a one-to-many scenario where nonces may become out of order and to avoid users needing to wait for a valid pending transaction to be mined before they may sign another transaction (nonces will be input but may be out of order):

// Maps a signer's address to the hash of a signature to a boolean indicating that signature's validity
mapping (address => mapping(bytes32 => bool)) public invalid_signatures;

// Functions only pass if the given signature has not been invalidated by the signer
modifier validSignature(address _signer, bytes _signature) {
    require(invalid_signatures[_signer][keccak256(_signature)] == false);
    _;
}    

/**
 * @param _signature: Signature to be revoked by the sender
*/
function revokeSignature(bytes _signature) public {
    invalid_signatures[msg.sender][keccak256(_signature)] = true;
}
```

**Comment by shanefontaine** [comment]
@bokkypoobah Are there any BTTS contracts in production on the main net right now? Have you found anything wrong with your proposal since the time of writing?

**Comment by bokkypoobah** [comment]
@shanefontaine There was a bug in v1.00 of the smart contract preventing the correct use of the `signedApproveAndCall(...)` function. This has now been fixed n v1.10. A v1.10 token contract in production - https://etherscan.io/address/0x4ac00f287f36a6aad655281fe1ca6798c9cb727b#code

**Comment by ptrwtts** [comment]
Nice work @wadeAlexC 

We actually came to the same conclusion around increase / decrease approval, and included them in our implementation that was submitted as a PR to Open Zeppelin here: https://github.com/OpenZeppelin/openzeppelin-solidity/pull/741. That code also includes a couple tweaks that were the result of an audit. Unfortunately we didn't update the spec here. Will try to get it updated so that we can work towards a common implementation.

In regards to the ability to revoke signatures, I agree that it seems valuable. I'll circle it internally to see if other's agree, and get an opinion on which implementation seems better.

One final thing that feels like an open question to me, and would be good to get a second opinion on, is whether a delegate should be explicitly defined in the presigned transaction. There was some discussion in the thread, but we never reached a consensus. On one hand, leaving it undefined allows anyone to pick up the transaction and submit it. But on the other hand, this opens it up to front running. My latest thinking is that a delegate should be explicitly specified. You could create a secondary fee market by having delegates first submit quotes, and then the user select someone. This seems better to me than a free-for-all where delegates are front running each other to claim the fee. What do others think?

**Comment by admazzola** [comment]
I think it could be valuable to have an explicit delegate address specified but make it optional. If it is 0x0 allow anyone to be a delegate?  The only downside then is feature creep

**Comment by gleissonassis** [comment]
I'm trying to use this pattern in my project but I'm getting error so I've checked if the recovered address from the signature was the same and I got different results. Is there something wrong with this piece of code? 

```javascript
transferPreSigned: function(from, to, amount) {
      var fee = 10;
      var chain = Promise.resolve();
      var nonce = 0;
      var hash = null;
      var signature = null;

      return chain
        .then(function(r) {
          nonce = 32;

          return token.methods.transferFromPreSignedHashing(
            token.options.address,
            from.address,
            to.address,
            amount,
            fee,
            nonce).call();
        })
        .then(function(r) {
          hash = r;
          console.log('Hash: ' + r);

          return web3.eth.accounts.sign(hash, from.privateKey);
        })
        .then(function(r) {
          signature = r.signature;
          console.log('Signature: ' + signature);

          return token.methods.recover(
            hash,
            signature).call();
        })
        .then(function(r) {
          console.log('Address (ori)', from.address);
          console.log('Address (rec)', r);
        });
    }
```

Variables sent to the function:
```javascript
var from = {
  address: '0xa933582bD31552b04790131dD885C2c7BEE0F0e5',
  privateKey: '0x0ef00d222ef18638bb5ba8bbbca54611d06aeabf26ffd2c9e8c5b760aa6c7bfe'
}

var to = {
  address: '0xf370872B4bAee32A73205b133D9Ad974F3AD437e',
  privateKey: '0x70a0b6f4bb138f3da385c39fd141d4615fdf1cd74a6fd6b590f4e2929d05ad72'
}

var amount = 100

token.options.address = '0x0DE841e872e246B59564F66953E078B59372dB2E'
```

I'm getting these results:
```
Hash: 0x49c58ae7dc453bd37ed678e032079abfa15d52cc4f97e3083c5008e5c02f7af6
Signature: 0x712e444e7c6f2c6c2d3f4144ad06cde80bafe35257016678908ecb29a21a6b3e4fa0f473df9d1f56da5dcc6563a8639d7d35eb0f17811c91825c03c80a9d93d91b
Address (ori) 0xa933582bD31552b04790131dD885C2c7BEE0F0e5
Address (rec) 0x139917E0C08a43c542Aa18e934a293533E0336B9
```
But if I try use this the values are the same:

```javascript
var sig = web3.eth.accounts.sign(hash, from.privateKey)
var rec = web3.eth.accounts.recover(hash, sig.signature);

console.log('Address (ori)', from.address);
console.log('Address (rec)', rec);
```

**Comment by RobertMCForster** [comment]
@gleissonassis This is likely because web3.eth.accounts.sign is using the [ERC191 signing prefix](https://github.com/ethereum/EIPs/issues/191) whereas the original Solidity is not (I assume the ethUtil that was used in the lgalabru implementation doesn't use the prefix). 

Check out the 3esmit and bokkypoobah implementations above to see what they do with the prefix. We've also got a full and audited token live at the [Coinvest GitHub](https://github.com/CoinvestHQ/COIN/blob/master/Contracts/Token/CoinvestTokenV2.sol) you can check out.

**Comment by shrugs** [comment]
uport's MetaTx from ages ago seems relevant here for some prior-art/context: https://github.com/uport-project/uport-identity/blob/develop/contracts/TxRelay.sol

**Comment by gleissonassis** [comment]
@RobertMCForster i've changed the recover method to add the prefix:

```solidity
      if (v != 27 && v != 28) {
        return (address(0));
      } else {
        bytes memory prefix = "\x19Ethereum Signed Message:\n32";
        bytes32 prefixedHash = keccak256(prefix, hash);
        return ecrecover(prefixedHash, v, r, s);
      }
```

And now is working fine, but I will read the code you suggested to find a better way to solve this issue.

Thanks!

**Comment by lavawallet** [comment]
Hello,

I improved upon this specification and have a wonderful working version live at http://lavawallet.io .  

Here is what was improved:

1. The relayer is not defined on-chain so that anyone can be a relay
2. There is a defined 'relayerReward' uint instead which defines a number of tokens which will be given from 'From' to 'msg.sender', whoever submits the tx to the mainnet and pays the gwei to do so
3. There is an expiration field so that the 'signed data' wont last forever, it is expressed in a block number 
4. There is a contractAddress field so that the signed data is locked to a specific contract which holds the tokens and which is conducting the logic for this system (called a Wallet Contract) 
5. There is a 'method' bytes field which specifies the method which is allowed to be remotely called by the relay with this signed offchain data: approve, transfer, withdraw, etc.

All of this data is signed using ECRecovery by Zeppelin 

Please see the contract here (yes it works):  0xf226b12c03514571c5a473b2627f5528da46d263

Thoughts ??? Should this be a different EIP or an improvement to this EIP ???

**Comment by PaulinJean** [comment]
var to = {
  address: '0x1846349d593c964B001Dc564fe38fc2099BB3b14',
  privateKey: 'ea3b591c158733c3c007f1cd6cc7e0455acc7ba72c5d1f0867de57e0e1585a59'
}

var amount = 100

token.options.address = '0x1846349d593c964B001Dc564fe38fc2099BB3b14'

**Comment by AusIV** [comment]
Is there any plan to align this with the finalized version of the EIP-712 spec? It matched a draft of EIP-712 from February, but the addition of a domain separator to the EIP-712 spec has broken the EIP-865 signature format, meaning there will likely be clients incapable of signing the required messages.

**Comment by tochie** [comment]
> Here's an implementation that is live - https://github.com/bokkypoobah/BokkyPooBahsTokenTeleportationServiceSmartContract/blob/master/contracts/BTTSTokenFactory.sol#L365-L395
> 
> See https://github.com/bokkypoobah/BokkyPooBahsTokenTeleportationServiceSmartContract#how-it-works

So can I use this and modify the contract name?

**Comment by bokkypoobah** [comment]
> > Here's an implementation that is live - https://github.com/bokkypoobah/BokkyPooBahsTokenTeleportationServiceSmartContract/blob/master/contracts/BTTSTokenFactory.sol#L365-L395
> > See https://github.com/bokkypoobah/BokkyPooBahsTokenTeleportationServiceSmartContract#how-it-works
> 
> So can I use this and modify the contract name?
It's MIT licensed, so please do.

If you don't want the hassles of deploying it yourself, there is a factory that deploys new token contracts @ https://etherscan.io/address/0x14aabc5ade82240330e5be05d8ef350661aebb8a#code . You can deploy you own token contract via the factory using a wallet GUI or some `geth` JavaScript console commands.


Note that this was written before ERC865 was proposed, and does not conform to the ERC865 standard (but works similarly).

**Comment by tochie** [comment]
@bokkypoobah Thanks very much.In order for me to learn the details I have decided to go the "deploy it yourself" route. I am reading the ERC865 version and trying to figure out how to go about it. I will let you (and anyone interested) know about my progress or if I encounter any issues. 

Thanks again.

**Comment by tochie** [comment]
@bokkypoobah I was going through https://bit.ly/2DrJGo5 and you mentioned: 

---
The token owner 0xa33a then uses their private key to sign the hash computed in the previous step using the function var sig = web3.eth.sign(**from**, signedTransferHash);
---

Where's the private key? I expected the 1st param to be the initiator's private key. No?

**Comment by lgalabru** [comment]
@tochie if you pre-sign your transfer using web3, you don't have to pass the private key, you just need to make sure that the account you want to sign with is unlocked (
https://web3js.readthedocs.io/en/1.0/web3-eth.html#sign)

**Comment by bokkypoobah** [comment]
> @bokkypoobah I was going through https://bit.ly/2DrJGo5 and you mentioned:
> 
> The token owner 0xa33a then uses their private key to sign the hash computed in the previous step using the function var sig = web3.eth.sign(**from**, signedTransferHash);
> 
> Where's the private key? I expected the 1st param to be the initiator's private key. No?

The `from` parameter will contain the **public** key. If you are using `geth`, you have to unlock your account using `personal.unlockAccount(fromPublicKey, "password")` . If you then call `var sig = web3.eth.sign(from, signedTransferHash);`, `geth` will sign your data using the unlocked **from** account private key

**Comment by tochie** [comment]
Thanks for the feedback.

@bokkypoobah @lgalbru I am not able to get past here: 

`var signedTransferHash = token.signedTransferHash(from, to, tokens, fee, nonce);`
I get a `TypeError: token.signedTransferHash is not a function`. In my implementation, `token` was set from `new web3.eth.Contract(ABI, tokenContractAddress)` where the `tokenContractAddress` variable is the address of the token I generated from the `BTTTokenFactory`. I also console logged `token`, and couldn't find `signedTransferHash`. Please advise. Thanks.

**Comment by tochie** [comment]
After a careful research guys, I realized that those functions where shoveled under `methods`. 👍

**Comment by tochie** [comment]
@bokkypoobah, @lgalabru 

Now here's my real roadblock: 

` Error: Invalid number of parameters for "signedTransfer". Got 8 expected 7!`

The tutorial from  https://bit.ly/2DrJGo5 has this test case: 

`var signedTransfer1Tx = token.signedTransfer(from, to, tokens, fee, nonce, sig, feeAccount, 
  {from: contractOwnerAccount, gas: 200000, gasPrice: defaultGasPrice});` which has 8 params. How were you able to succeed here.

Please advise. Thanks.

**Comment by RobertMCForster** [comment]
@tochie That call has 7 parameters then the default transaction parameters. The only parameters for the function are:
 `from, to, tokens, fee, nonce, sig, feeAccount`
Check to make sure your formatting is correct and there are no parameters other than those.

**Comment by tochie** [comment]
@RobertMCForster Yes. I know there are 7 params but the examples from the tutorials and authors notes passed in 8 param [the default 7 params and an extra object that captures the contract owners address].

**Comment by tochie** [comment]
Anyone to answer me? My entire application is dependent on this. For now, its a massive roadblock :)

**Comment by AFDudley** [comment]
The fact that no one has mentioned https://github.com/ethereum/EIPs/issues/28 in this thread is somewhat alarming. 🙂

**Comment by tbocek** [comment]
There is a signature malleability issue in the full implemention mentioned on top, see https://github.com/OpenZeppelin/openzeppelin-solidity/pull/1622

**Comment by ptrwtts** [comment]
Hi @RobertMCForster, thanks for sharing your solution.

What are your thoughts on what @tbocek proposed, where you only allow 27/28 signatures, and force clients to use them: https://github.com/tbocek/openzeppelin-solidity/commit/dc168f3bf6608cdbc3f51f2dda4560815afd482c

With this, it should still be safe to use the signature for uniqueness. One advantage is that you don't need to use incrementing nonces and store them on chain, which complicates signature generation (need to lookup the last nonce) and costs more gas. You can just use something random (e.g. current timestamp) as the nonce. 

The downside of this approach seems to be that not all clients will be compatible, if they don't adhere to only generating 27/28 signatures.

Let me know if I'm missing something.

**Comment by RobertMCForster** [comment]
@ptrwtts , @tbocek 's solution is still vulnerable to signature malleability. 

I actually missed that his post didn't mention the other, more traditional, malleability vulnerability that also comes from this method of using signatures as unique identifiers.

The problem is that with ECDSA each signature has an inverse that will also recover correctly. 

Per Vitalik:
> Allowing transactions with any s value with 0 < s < secp256k1n, as is currently the case, opens a transaction malleability concern, as one can take any transaction, flip the s value from s to secp256k1n - s, flip the v value (27 -> 28, 28 -> 27), and the resulting signature would still be valid.

That quote is taken from the Homestead hardfork EIP [here](https://eips.ethereum.org/EIPS/eip-2), which also explains that they left the validity of the top half of the elliptic curve in the ecrecover pre-compiled contract in order to maintain backwards compatibility with old Bitcoin signatures.

Now, you can restrict it even further to allow, say, only 27, but at that point you're just complicating things for no reason. The transaction hash (maybe it is better termed "parameter hash" to differentiate it from an Ethereum transaction hash) doesn't need incrementing nonces either as long as nonce is unique, it's inherently compatible with all clients, and it's just a much safer way to do things.

**Comment by lgalabru** [comment]
@RobertMCForster @tbocek @ptrwtts  thank you guys for the heads up!
What's your take on the edit?

**Comment by tbocek** [comment]
@RobertMCForster my solution is not vulnerable to signature malleability. The [spec](https://ethereum.github.io/yellowpaper/paper.pdf) says in Appendix F, that s is only valid with values 0< s <secp256k1n÷2 + 1 (281). Maybe are referring to an old version. The spec also says that only 27/28 is allowed (282). In [EIP-2](https://eips.ethereum.org/EIPS/eip-2) chapter specification, point 2, they did not left the top half for backwards compatibility:

> All transaction signatures whose s-value is greater than secp256k1n/2 are now considered invalid. 

By restricting to 27/28 you make it according to the specification, and fix the signature malleability. Otherwise, ERC865 as it is currently implemented can be exploited and can even potentially loose coins or tokens.

**Comment by tbocek** [comment]
@ptrwtts I think the nonce adds little value to the security. Or maybe I missed the reason why this nonce is important in the current implementation.

**Comment by RobertMCForster** [comment]
@tbocek The fixes you're referring to were only applied on the Ethereum network itself, not the ecrecover pre-compiled contract.

From that EIP-2 link:
> All transaction signatures whose s-value is greater than secp256k1n/2 are now considered invalid. **The ECDSA recover precompiled contract remains unchanged and will keep accepting high s-values; this is useful e.g. if a contract recovers old Bitcoin signatures.** [bold mine]

To test it yourself, sign some data, take the s value, flip it with this: `s2 = 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141 - s1;`, then flip the v from 27 -> 28 or vice versa. You'll see that both signatures recover correctly on-chain.

ERC865 _was_ able to be exploited and _could_ lose users coins. We actually had to hardfork our token because we used signatures as unique identifiers and, while we caught the 0/1 or 27/28 vulnerability before launch, we missed the complementary signature problem.

**Comment by AusIV** [comment]
I believe that in the current implementation, the nonce simply allows you to send the same quantity of the same token to the same recipient multiple times. Absent the nonce, you would have the same message and produce the same signature.

**Comment by RobertMCForster** [comment]
@AusIV How would it allow this? The contract requires the hashedTx has not been used then declares it used. If the nonce is the same twice the transaction will revert on `require(hashedTxs[hashedTx] == false);`

**Comment by tbocek** [comment]
@RobertMCForster you are right. Thanks for clarification!

**Comment by tbocek** [comment]
@RobertMCForster What was your solution. Did you normalize the signature, something like:

```
if(s > halfOrder) {
  s = uint256(0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141) - s;
}
bytes memory normalizedSignature = abi.encode(r, s);
```
Then it can be check for the normalizedSignature, which is unique.

**Comment by RobertMCForster** [comment]
> 
> 
> @RobertMCForster What was your solution. Did you normalize the signature, something like:
> 
> ```
> if(s > halfOrder) {
>   s = uint256(0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141) - s;
> }
> bytes memory normalizedSignature = abi.encode(r, s);
> ```
> 
> Then it can be check for the normalizedSignature, which is unique.
> 
> Edit: I'm even considering to simply add a check (besides 0/1) to allow only signatures with a low order in Zeppelin's recover() function.

@tbocek My full solution is [here](https://github.com/CoinvestHQ/COIN/blob/master/contracts/CoinvestToken.sol). I used the hash of given parameters (including a random nonce) to keep track of unique transactions.

Disclaimers regarding the code: 
1. If written again I would remove the nonce mapping, nonce++, and getNonce() since, as @ptrwtts mentioned, a random value such as timestamp works fine (we're just using timestamp anyway on our off-chain code for simplicity). 
2. If you want the receiveApproval function to allow other ERC865's to send your tokens and pay their own gas, make sure you require msg.sender isn't address(this) or people can take stuck tokens from it. Oops.
3. There are random little gas efficiencies that could be done like not nesting transfer balance requires--I was paranoid from the bug on our first ERC865 version--, making gas cost calculations more accurate, and maybe do something like add amount to receiver, calculate gas, subtract total (amount+fee) from sender, add fee to delegate, instead of a full transfer to receiver then a full transfer to delegate.

**Comment by villesundell** [comment]
Is it intentional, that r, s and v are in "non-standard" order in the reference implementation:
https://github.com/OpenZeppelin/openzeppelin-solidity/pull/741/files#r264813474 ?

**Comment by villesundell** [comment]
It seems that it's intentional: same approach is used by the Zeppelin recover() implementation (https://github.com/OpenZeppelin/openzeppelin-solidity/blob/v1.12.0/contracts/ECRecovery.sol) because of web3.eth.sign().

**Comment by 0xbitcoin** [comment]
I deployed a working implementation of this complete with open source relay
servers!!! Check it out :

Relay.0xbtc.io
Lavatoken.com

There is also a fully-featured ERC20 Token wallet that supports this that
is also open source called LAVA wallet.

  It uses Signed Typed Data to allow users to submit an offchain signature
to any relay with an embedded incentivization reward in tokens.   When they
relay server submits the tx data and it's offchain signature to the
contract on mainnet, it has to pay gas but the users tokens are transferred
to the users recipient and the relay server gets the token incentive reward
!! It works great 😊

Infernaltoast

**Comment by oberstet** [comment]
As far as I can see, there are currently 2 proposals for tokens that allow gasless token transfers:

* EIP #865
* EIP #965

Are there more proposals?
Any new directions or recommendations rgd this feature in general?

**Comment by lastmjs** [comment]
Hey everyone, I'm new to this, but this feature is very important to a use case for my project. What is involved in getting this proposal or similar proposals for gas delegation moving and implemented, and how long should I expect this to take? I'm most likely willing to put significant effort into helping this along come summer, I just need some guidance.

**Comment by embedovator** [comment]
Hello everyone! Total n00b here.

I’d like to add a use case that may be of interest: IoT devices. I haven't seen such a use case mentioned here (nor in EIPs #965 and #1776). Currently, IoT devices that want to interact with the network constantly need "refilling" of gas by their respective owners. The device may request that the owner give it more gas, or the burden can be put on the owner to keep track of gas usage. 

However, if a third party were to pay for the gas, that would make future network interactions much less cumbersome. On behalf of many future IoT devices, ERC #865 seems like it would be real useful!

**Comment by chompomonim** [comment]
Actually, from my point of view, IoT devices should use 2nd layer platforms.

**Comment by embedovator** [comment]
> Actually, from my point of view, IoT devices should use 2nd layer platforms.

Interesting! Thank you for introducing me to 2nd layer platforms. I found this [definition](https://www.ethnews.com/a-survey-of-second-layer-solutions-for-blockchain-scaling-part-1):

> Simply put, the second layer refers to all the auxiliary platforms and protocols that act as digital dumping grounds where activities can be offloaded from main chains to save storage space.

I completely agree that we don't want a bunch of tiny devices dumping data on the main blockchain, i.e. base-layer. I suppose that's been shown to be unscalable, a lá CryptoKitties. I wonder if there are use cases for any IoT devices to post data to the main chain? 

I did some more research into other 2nd layer platforms and found this helpful [article](https://www.zerohedge.com/news/2019-03-31/implications-fusing-5g-and-blockchain), which describes the scaling problem while also suggesting to use 2nd layer platforms as @chompomonim had mentioned. 

Apparently [Plasma](https://plasma.io) was going to be a 2nd layer platform, but was [halted](https://www.newsbtc.com/2018/11/09/plasma-halted-as-ethereum-seeks-alternative-scaling-in-zk-snarks/). Pardon my not being current on the news. As of March 2019, [here](https://media.consensys.net/the-state-of-ethereum-layer-2-protocol-development-2-f22b2603abd6) is the "State of the Ethereum Layer 2 Protocol Development", which states:
> The two main scaling proposals on Ethereum for the past couple years have been child chains (i.e. Plasma) and state channels (more general versions of payment channels, which have been around for many years on the Bitcoin network).

Ok, great! Thank you for the introduction to these 2nd layer platforms. I'll do more research into them.

**Comment by 1blockologist** [comment]
@embedovator ERC #777 has delegates which allows the issuer to pay for gas. So there is topping up but only by the company, which should be done sparingly based on the use case of course so that it isn't prohibitively expensive, but it removes the burden from the client and makes it seem like a gas-less token transfer. Just food for thought.

**Comment by GiacomoPT** [comment]
can i use this implementation already on an existing ERC20?

**Comment by ytrezq** [comment]
Just a question, why not allowing buying gas with token instead of Ether ? That would be much simpler, isn’t it ?

**Comment by chompomonim** [comment]
@ytrezq that is very hard because miners would need to know how to deterministically price tokens. Also even if price problem would be solved, that anyway would mean hardfork and a lot of changes in ethereum node implementations.

Such pricing of gas with tokens instead of Ether most probably will never happen.

**Comment by ytrezq** [comment]
@chompomonim : I was pointing that this is still simpler than paying transactions/computations directly in tokens, isn’t it ?

**Comment by chompomonim** [comment]
@ytrezq  what do you mean by **"this"**?

**Comment by ytrezq** [comment]
@chompomonim : wanted to mean `it’s`. I just used the wrong word.

**Comment by chompomonim** [comment]
I mean any such changes on protocol level are very hard and almost impossible. So only workaround such as this and similar are possible.

**Comment by erc20devteam** [comment]
Hi there, anyone here who have deployed ERC865 on main net and also that token is listed on active exchanges and on coinmarketcap or coingecko. Please share or post the Etherscan Link. Thanks.

**Comment by SirPhemmiey** [comment]
Hello guys, has this proposal been approved?

**Comment by chompomonim** [comment]
@SirPhemmiey No. And there is very little chance that it will be ever approved.

**Comment by SirPhemmiey** [comment]
Oh oh. If I may ask, why?

**Comment by 1blockologist** [comment]
> Oh oh. If I may ask, why?

Think about it.

It requires miners to collectively agree that native Ether has no use case anymore.

**Comment by admazzola** [comment]
This is very much untrue. Ether is used for staking and deploying contracts
regardless.   This is just removing friction and making user experience
better.

**Comment by 1blockologist** [comment]
saying "no use case" was hyperbole, obviously, everyone in this thread knows it will reduce friction, you have to convince the miners with their predictable rebuttals.

**Comment by 3esmit** [comment]
@1blockologist No, it dont requires miners to agree on anything. This is an application level interface.

**Comment by 1blockologist** [comment]
> @1blockologist No, it dont requires miners to agree on anything. This is an application level interface.

then what is preventing it from moving forward? the author's neglect?

**Comment by 3esmit** [comment]
The way it implemented seems not attractive to developers because it is limited to transfers, cannot do general executions on Ethereum Smart Contracts. 

There are other proposals, like the draft in https://eips.ethereum.org/EIPS/eip-1077 that work on top of account contracts, as in this example here https://github.com/status-im/account-contracts I am implementing this EIP1077 (alongside my custom gasRelay interface that I plan to use as an update in 1077). 
1077 is also being headed by Universal Logins https://universallogin.io/ however to 1077 interface is the last thing we would probably finalize, as we are still tweaking the ideal way of refunding the gas in the smart contract in account contracts.

ERC865 works on updating and ERC20 token, so it would only work on new tokens that implement this interface, while 1077 proposes a method that any (valued) fungible token is supported, and can do anything (transfer token, transfer other token, create a contract, execute anything that a smart contract allows), the limitation being that account contracts are supported by the DApp (some might not be, like the ones using ethereum signed messages and not supporting ERC1271 interface)

The Ether remains part of it, unless miners/validators soft-fork to they become relayers themselves. Ether would still be more valuable as a native token payment is cheaper then an ERC20 token, and also allows Direct execution, while ERC20 relays on the assumption that someone will relay that for you.

**Comment by charlesvien** [comment]
👍

**Comment by Adicz345** [comment]
975413

**Comment by MicahZoltu** [comment]
@alita-moore Do you happen to know why the stale bot seems to be blind to this issue?

----

I'm closing this issue as it hasn't had any action for quite some time.  If someone would like to pursue this idea further, please create a thread over at Ethereum Magicians forum to discuss further, or create an EIP by following the process laid out in [EIP-1](https://eips.ethereum.org/EIPS/eip-1).
